File: src/utils/date.ts
export class CustomDate {
  public date: Date;

  public constructor() {
    this.date = new Date();
  }

  public year() {
    return this.date.getFullYear();
  }

  public month() {
    return this.date.getMonth() + 1;
  }

  public day() {
    return this.date.getDate();
  }

  /**
   *
   * @returns `MM`
   */
  public padMonth() {
    return String(this.month()).padStart(2, "0");
  }

  /**
   *
   * @returns `dd`
   */
  public padDay() {
    return String(this.day()).padStart(2, "0");
  }

  /**
   *
   * @returns `YYYY-MM-dd`
   */
  public input() {
    return `${this.year()}-${this.padMonth()}-${this.padDay()}`;
  }

  /**
   * 
   * @param value `d-M-YYYY`
   * @returns `YYYY-MM-dd`
   */
  static to_input(value: string) {
    const dateRegex = new RegExp(/(?<year>\d{4})-(?<month>\d+)-(?<day>\d+)/);
    
    const result = dateRegex.exec(value)?.groups;

    if (!result) return new CustomDate().input();


    const year = result.year;
    const month = result.month.padStart(2, "0");
    const day = result.day.padStart(2, "0");

    return `${year}-${month}-${day}`;
  }
}

File: src-tauri/src/libs/date.rs
#![allow(dead_code)]

pub struct CustomDate {
  day: usize,
  month: usize,
  year: usize,
}

impl CustomDate {
  pub fn new(day: usize, month: usize, year: usize) -> Self {
    Self { day, month, year }
  }

  pub fn year(&self) -> usize {
    self.year
  }

  /// Pad Day
  /// 
  /// Padded Day
  /// 
  /// input: 5, output: 05,
  /// input: 11, output: 11,
  pub fn pad_day(&self) -> String {
    pad_usize(self.day, 2, "0")
  }

  /// Pad Month
  /// 
  /// Padded month
  /// 
  /// input: 5, output: 05,
  /// input: 11, output: 11,
  pub fn pad_month(&self) -> String {
    pad_usize(self.month, 2, "0")
  }
}

fn pad_usize(value: usize, len: usize, with: &str) -> String {
  let mut value = format!("{value}");

  if value.len() < len {
    let mut padding = String::new();

    for _i in value.len()..len {
      padding.push_str(&with);
    }
    value =  format!("{padding}{value}");

    value
  } else {
    value
  }
}